pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,967 @@
+	/// amount.
+	///
+	/// Example:
+	/// * Say before my pre invest state has `foreign_done = 1000` and
+	/// `foreign_swap.amount = 500`. Now we look at three scenarios in which we
+	/// increase below, exactly at and above the `foreign_swap.amount`:
+	/// * a) If we increase by 500, we can reduce the `foreign_swap.amount`
+	///   fully, which we denote by adding the 500 to the `foreign_done` amount.
+	///   Moreover, we can immediately invest the 500. The resulting state is
+	///   `(done_amount = 1500, investing = 500)`.
+	/// * b) If we increase by 400, we can reduce the `foreign_swap.amount` only
+	///   by 400 and increase both the `investing` as well as `foreign_done`
+	///   amount by that. The resulting state is
+	/// `(done_amount = 1400, foreign_swap.amount = 100, investing = 400)`.
+	/// * c) If we increase by 600, we can reduce the `foreign_swap.amount`
+	///   fully and need to add a swap into pool currency for 100. Moreover both
+	///   the `investing` as well as `foreign_done` amount can only be increased
+	///   by 500. The resulting state is
+	/// `(done_amount = 1500, pool_swap.amount = 100, investing = 500)`.
+	///
+	/// NOTE: We can ignore handling all states which include
+	/// `*SwapIntoForeignDone` without `ActiveSwapIntoForeignCurrency*` as we
+	/// consume the done amount and transition in the post transition phase.
+	/// To be safe and to not make any unhandled assumptions, we throw
+	/// `DispatchError::Other` for these states though we need to make sure
+	/// this can never occur!
+	fn handle_increase(&self, swap: Swap<Balance, Currency>) -> Result<Self, DispatchError> {
+		if swap.currency_in == swap.currency_out {
+			return Self::handle_increase_non_foreign(self, swap);
+		}
+
+		match &self {
+			Self::NoState => Ok(Self::ActiveSwapIntoPoolCurrency { swap }),
+			// Add pool swap
+			Self::InvestmentOngoing { invest_amount } => {
+				Ok(Self::ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+					swap,
+					invest_amount: *invest_amount,
+				})
+			}
+			// Bump pool swap
+			Self::ActiveSwapIntoPoolCurrency { swap: pool_swap } => {
+				swap.ensure_currencies_match(pool_swap, true)?;
+				Ok(Self::ActiveSwapIntoPoolCurrency {
+					swap: Swap {
+						amount: swap.amount.ensure_add(pool_swap.amount)?,
+						..swap
+					},
+				})
+			}
+			// Reduce foreign swap amount by the increasing amount and increase investing amount as
+			// well adding foreign_done amount by the minimum of active swap amounts
+			Self::ActiveSwapIntoForeignCurrency { swap: foreign_swap } => {
+				swap.ensure_currencies_match(foreign_swap, false)?;
+				let invest_amount = swap.amount.min(foreign_swap.amount);
+				let done_amount = swap.amount.min(foreign_swap.amount);
+
+				match swap.amount.cmp(&foreign_swap.amount) {
+					// pool swap amount is immediately invested and done amount increased equally
+					Ordering::Equal => {
+						Ok(
+							Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
```

Comment:
Not sure why increasing an `ActiveSwapIntoPoolCurrency` state adds an `InvestmentOngoing` to the resulting state